src/drt/src/frDesign.h

```diff
@@ -98,20 +101,38 @@ class frDesign
                                                 !isVerticalLayer(layerNum));
   }
 
+  void addUpdate(const drUpdate& update)
+  {
+    if (updates_.size() == 0)
+      updates_.resize(MAX_THREADS * 2);
+    auto num_batches = updates_.size();
+    updates_[updates_sz_++ % num_batches].push_back(update);
+  }
+  const std::vector<std::vector<drUpdate>>& getUpdates() const
+  {
+    return updates_;
+  }
+  bool hasUpdates() const { return updates_sz_ != 0; }
+  void clearUpdates()
+  {
+    updates_.clear();
+    updates_sz_ = 0;
+  }
+  void incrementVersion() { ++version_; }
+  int getVersion() const { return version_; }
+
+  ~frDesign() {}
+
  private:
   std::unique_ptr<frBlock> topBlock_;
   std::map<frString, frMaster*> name2master_;
   std::vector<std::unique_ptr<frMaster>> masters_;
   std::unique_ptr<frTechObject> tech_;
   std::unique_ptr<frRegionQuery> rq_;
+  std::vector<std::vector<drUpdate>> updates_;
+  int updates_sz_;
```

Comment:
Why do we store updates_sz separately from update_.size() ?

Reply:
As you can see updates_ is a vector of vector of updates. The outer vector size is MAX_THREADS*2 (fixed). This is to serialize each of the inner vectors in parallel (To speedup updates serialization).
Another option would be to have a normal 1d vector<drUpdate> and do the splitting latter at serialization. Tell me if you feel the second option is better.

Reply:
The outer and inner vectors both know their size so I still don't get why you need to store it separately.